File: app/utils/array.lua
local ____lualib = require("lualib_bundle")
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArrayReduce = ____lualib.__TS__ArrayReduce
local __TS__ArrayFindIndex = ____lualib.__TS__ArrayFindIndex
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local __TS__ArrayReverse = ____lualib.__TS__ArrayReverse
local __TS__ArrayConcat = ____lualib.__TS__ArrayConcat
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local ____exports = {}
____exports.array = {}
local array = ____exports.array
do
    local _popEl
    function array.iterate_back(arr, handler)
        do
            local i = #arr - 1
            while i >= 0 do
                handler(arr[i + 1], i)
                i = i - 1
            end
        end
    end
    function _popEl(arr, idx)
        local element = arr[idx + 1]
        __TS__ArraySplice(arr, idx, 1)
        return element
    end
    function array.random_element(arr)
        local ____temp_0
        if #arr > 1 then
            ____temp_0 = arr[math.random(0, #arr - 1) + 1]
        else
            ____temp_0 = arr[1]
        end
        return ____temp_0
    end
    function array.clear(arr, callback_befor_remove)
        if callback_befor_remove then
            for ____, el in ipairs(arr) do
                callback_befor_remove(el)
            end
        end
        __TS__ArraySplice(arr, 0, #arr)
    end
    function array.sum(arr)
        return #arr > 1 and __TS__ArrayReduce(
            arr,
            function(____, acc, cur) return acc + cur end,
            0
        ) or arr[1]
    end
    function array.remove(arr, par)
        if type(par) == "function" then
            local index = __TS__ArrayFindIndex(
                arr,
                function(____, v) return par(v) end
            )
            if index ~= -1 then
                __TS__ArraySplice(arr, index, 1)
            end
        else
            local index = __TS__ArrayIndexOf(arr, par)
            if index ~= -1 then
                __TS__ArraySplice(arr, index, 1)
            end
        end
    end
    function array.removeIndex(arr, index)
        if index ~= -1 then
            __TS__ArraySplice(arr, index, 1)
        end
    end
    function array.removeIndexes(arr, indexes)
        array.iterate_back(
            indexes,
            function(i) return __TS__ArraySplice(arr, i, 1) end
        )
    end
    function array.exclude(arr, condition)
        return __TS__ArrayFilter(
            arr,
            function(____, el) return not condition(el) end
        )
    end
    function array.pop(arr, idx)
        local ____temp_1
        if idx == nil then
            ____temp_1 = table.remove(arr)
        else
            ____temp_1 = _popEl(arr, idx)
        end
        local element = ____temp_1
        return element
    end
    function array.first(arr)
        return arr[1]
    end
    function array.last(arr)
        local ____temp_2
        if #arr > 1 then
            ____temp_2 = arr[#arr]
        else
            ____temp_2 = array.first(arr)
        end
        return ____temp_2
    end
    function array.shuffle(arr)
        math.randomseed(os.time())
        do
            local i = #arr - 1
            while i > 0 do
                local j = math.floor(math.random() * (i + 1))
                local ____temp_3 = {arr[j + 1], arr[i + 1]}
                arr[i + 1] = ____temp_3[1]
                arr[j + 1] = ____temp_3[2]
                i = i - 1
            end
        end
    end
    function array.inverse(arr)
        return __TS__ArrayReverse(arr)
    end
    --- Добавляет элемент в начало массива, возвращая новый массив
    function array.push_begin(arr, el)
        return __TS__ArrayConcat({el}, arr)
    end
    function array.indexes(arr)
        return __TS__ArrayMap(
            arr,
            function(____, e, i) return i end
        )
    end
end
return ____exports

File: src/utils/array.ts
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable no-inner-declarations */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace array {

    export function random_element<T>(arr: T[]): T {
        return arr.length > 1 ? arr[math.random(0, arr.length - 1)] : arr[0];
    }
    export function clear<T>(arr: T[], callback_befor_remove?: (element: T) => void) {
        if (callback_befor_remove)
            for (const el of arr)
                callback_befor_remove(el);
        arr.splice(0, arr.length);
    }

    export function sum(this: void, arr: number[]): number {
        return arr.length > 1 ? arr.reduce((acc, cur) => acc + cur, 0) : arr[0];
    }

    export function remove<T>(arr: T[], condition: ((element_in_arr: T) => boolean)): void;
    export function remove<T>(arr: T[], element: T): void;
    export function remove<T>(arr: T[], par: ((element_in_arr: T) => boolean) | T) {
        if (typeof (par) === 'function') {
            const index = arr.findIndex((v) => (par as any)(v));
            if (index != -1)
                arr.splice(index, 1);
        }
        else {
            const index = arr.indexOf(par);
            if (index != -1)
                arr.splice(index, 1);
        }

    }

    export function removeIndex<T>(arr: T[], index: number) {
        if (index != -1)
            arr.splice(index, 1);
    }
    export function removeIndexes<T>(arr: T[], indexes: number[]) {
        iterate_back(indexes, (i) => arr.splice(i, 1));
    }
    export function exclude<T>(arr: T[], condition: (element: T) => boolean) {
        return arr.filter(el => !condition(el));
    }
    export function pop<T>(arr: T[], idx?: number) {
        const element = idx == undefined ? arr.pop() : _popEl(arr, idx);
        return element;
    }


    export function first<T>(arr: T[]): T { return arr[0]; }
    export function last<T>(arr: T[]): T { return arr.length > 1 ? arr[arr.length - 1] : first(arr); }

    export function shuffle<T>(arr: T[]): void {
        math.randomseed(os.time());
        for (let i = arr.length - 1; i > 0; i--) {
            const j = math.floor(math.random() * (i + 1));
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }

    export function inverse<T>(arr: T[]): T[] { return arr.reverse(); }

    export function iterate_back<T>(arr: T[], handler: (el: T, i: number) => void) {
        for (let i = arr.length - 1; i >= 0; i--)
            handler(arr[i], i);
    }

    /**Добавляет элемент в начало массива, возвращая новый массив */
    export function push_begin<T>(arr: T[], el: T) {
        return [el].concat(arr);
    }

    function _popEl<T>(arr: T[], idx: number): T {
        const element = arr[idx];
        arr.splice(idx, 1);
        return element;
    }

    export function indexes<T>(arr: T[]): number[] {
        return arr.map((e, i) => i);
    }
}

